### PR TITLE
Add CLI --help option

### PR DESCRIPTION
**Pull request by grodin**
Intended to land on top of #476

### CODE REVIEW

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 1]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
...
+      |     $ ktfmt --dropbox-style Main.kt src/Parser.kt
+      |     Done formatting Main.kt
+      |     Error formatting src/Parser.kt: @@@ERROR@@@; skipping.
+      |    
+      |Commands options:
+      |  -n, --dry-run                     Don't write to files, only report files which 
+      |                                        would have changed
+      |  --style=dropbox                   Use 4-space block indenting
+      |  --style=google                    Google style
+      |  --style=kotlinlang                Kotlin language guidelines style
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:75]: Thoughts?
```
--style=kotlinlang | google | dropbox   style for formatting approach
                                        kotlinlang - https://kotlinlang.org/docs/coding-conventions.html
                                        google     - add link
                                        dropbox    - Use 4-space block indenting
```

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt [hunk 2]
```diff
@@ -74,48 +78,50 @@ class Main(
     val parsedArgs =
         when (processArgs) {
           is ParseResult.Ok -> processArgs.parsedValue
-          is ParseResult.Error -> exitFatal(processArgs.errorMessage, 1)
+          is ParseResult.ShowMessage -> {
+            err.println(processArgs.message)
+            return EXIT_CODE_SUCCESS
+          }
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/Main.kt:84]: 🫶

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 3]
```diff
@@ -57,6 +101,9 @@ data class ParsedArgs(
       var removeUnusedImports = true
       var stdinName: String? = null
 
+      if (args.contains("--help"))
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:104]:
```
"--help" in args || "-h" in args
```

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 4]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
...
+    val HELP_TEXT = """ktfmt - command line Kotlin source code pretty-printer
+      |
+      |Usage:
+      |  ktfmt [--style=dropbox | --style=google | --style=kotlinlang ] [-n | --dry-run] [--set-exit-if-changed] 
+      |        [--do-not-remove-unused-imports] <File1.kt> <File2.kt> ...
+      |  ktfmt @ARGFILE
+      |
+      |      ktfmt formats Kotlin source code files in-place, reporting for each file whether the
+      |      formatting succeeded or failed on standard error. If none of the style options are
+      |passed, Facebook's style is used.
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:60]: Let's put Meta here instead. I plan on making that an explicit style. That way we can state that it is the default
`s/Facebook/Meta/`

**Review by hick209 — COMMENTED** [review]
I like the direction and honestly this could happen before we land the `--style` change as adding an option does not break the API.

Your call though

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 5]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
       return parseOptions(arguments)
     }
 
+    val HELP_TEXT = """ktfmt - command line Kotlin source code pretty-printer
+      |
+      |Usage:
+      |  ktfmt [--style=dropbox | --style=google | --style=kotlinlang ] [-n | --dry-run] [--set-exit-if-changed] 
+      |        [--do-not-remove-unused-imports] <File1.kt> <File2.kt> ...
+      |  ktfmt @ARGFILE
```
[the remark this answers is not in the document]
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:56]: Good catch, the `style=` change hasn't landed yet so this is incorrect anyway until then.

Some GNU tools (cp, mv, rsync) list their calling methods like this:

```
Usage: cp [OPTION] ...
  or:  cp [OPTION] ...
  or:  cp [OPTION] ...
```

and others (ripgrep, jq, bash) just list them similar to this:
```
Usage:
    rg [OPTIONS] ...
    rg [OPTIONS] ...
```

I think we should go for one or the other. I slightly prefer the second, ripgrep/jq style.

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 1 above)
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:75]: Again, this should have been the prior `--dropbox-style` etc. prior to the change.

I like your suggestion for when that lands though!

Reverted in 6dec31543cfb688627fdbe43600b5adc0709f055

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 4 above)
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:60]: 69f879a7931c598ccc16382856406648f0927e5c

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 3 above)
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt]: 69f879a7931c598ccc16382856406648f0927e5c

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 4 above)
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt]: 2573612d4091d90dc4e030fdb0be2cb3a66d38fe

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (hick209) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:56]: Good idea to follow what other well know CLI programs do.

Make sure to update the `style=` here and I can import this PR internally

**Review by hick209 — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (hick209) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:56]: Should we add something like "for more info use --help" in the general `USAGE` message? WDYT?

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt [hunk 6]
```diff
@@ -80,6 +80,10 @@ class Main(
     val parsedArgs =
         when (processArgs) {
           is ParseResult.Ok -> processArgs.parsedValue
+          is ParseResult.ShowMessage -> {
+            err.println(processArgs.message)
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/Main.kt:84]: Why `err`? Shouldn't it be `out` here?

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 7]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
       return parseOptions(arguments)
     }
 
+    val HELP_TEXT = """ktfmt - command line Kotlin source code pretty-printer
+      |
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:52]:
```kotlin
    val HELP_TEXT = """
      |ktfmt - command line Kotlin source code pretty-printer
      |
```

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 8]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
...
+      |ktfmt formats Kotlin source code files in-place, reporting for each file whether the
+      |formatting succeeded or failed on standard error. If none of the style options are
+      |passed, Meta's style is used.
+      |
+      |Alternatively, ktfmt can read Kotlin source code from standard in and write the 
+      |formatted result on standard output.
+      |
+      |Example:
+      |     $ ktfmt --dropbox-style Main.kt src/Parser.kt
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:66]: Need to update now that dropbox style is gone
```kotlin
      |     $ ktfmt --kotlinlang-style Main.kt src/Parser.kt
```

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt (hunk 6 above)
Comment (grodin) [reply core/src/main/java/com/facebook/ktfmt/cli/Main.kt:84]: Yep, you're right.

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 9]
```diff
@@ -48,6 +48,50 @@ data class ParsedArgs(
...
+      |     Done formatting Main.kt
+      |     Error formatting src/Parser.kt: @@@ERROR@@@; skipping.
+      |    
+      |Commands options:
+      |  -n, --dry-run                     Don't write to files, only report files which 
+      |                                        would have changed
+      |  --dropbox-style                   Use 4-space block indenting
+      |  --google-style                    Google style
+      |  --kotlinlang-style                Kotlin language guidelines style
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:75]: 1. Add `help` option
2. remove dropbox style
3. add meta style and mention it is the default
4. don't forget the `stdin-name` param

```kotlin
      |Commands options:
      |  -h, --help                        Show this help message
      |  -n, --dry-run                     Don't write to files, only report files which 
      |                                        would have changed
      |  --meta-style                      Use 2-space block indenting (default)
      |  --google-style                    Google internal style (2 spaces)
      |  --kotlinlang-style                Kotlin language guidelines style (4 spaces)
      |  --stdin-name=<name>               Name to report when formatting code from stdin
```

**Comment by hick209** [comment]
Which might be the last before 1.0.0

**Review by hick209 — APPROVED** [review]

**Comment by hick209** [comment]
Merging...

**Pull request closed without merging**